streams/src/main/java/org/apache/kafka/streams/processor/internals/RecordCollectorImpl.java

```diff
@@ -98,16 +95,14 @@ public RecordCollectorImpl(final Producer<byte[], byte[]> producer, final String
                     @Override
                     public void onCompletion(final RecordMetadata metadata, final Exception exception) {
                         if (exception == null) {
-                            if (sendException != null) {
-                                return;
-                            }
                             final TopicPartition tp = new TopicPartition(metadata.topic(), metadata.partition());
                             offsets.put(tp, metadata.offset());
                         } else {
-                            if (sendException == null) {
-                                sendException = exception;
-                                log.error("{} Error sending record to topic {}. No more offsets will be recorded for this task and the exception will eventually be thrown", logPrefix, topic, exception);
-                            }
+                            log.error("{} Error sending record with key {} to topic {}, partition {}. " +
+                                    "No more offsets will be " +
+                                    "recorded for this task and the exception will eventually be thrown",
+                                logPrefix, key, topic, partition, exception);
+                            throw new StreamsException(String.format("%s exception caught when producing", logPrefix), exception);
```

Comment:
The original approach is to avoid throwing exceptions on each of the record: for example, if you get a timeout exception on the request, all records in the batch will return the same exception in that callback, which will spill the log4j since we will get one error for each record.